ext/random/engine_xoshiro256starstar.c

```diff
@@ -0,0 +1,249 @@
+   |                     Takuji Nishimura                                 |
+   |                     Shawn Cokus <Cokus@math.washington.edu>          |
+   |                     David Blackman                                   |
+   |                     Sebastiano Vigna <vigna@acm.org>                 |
+   |                     Melissa O'Neill <oneill@pcg-random.org>          |
+   +----------------------------------------------------------------------+
+*/
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include "php.h"
+#include "php_random.h"
+
+#include "ext/spl/spl_exceptions.h"
+#include "Zend/zend_exceptions.h"
+
+static inline uint64_t splitmix64(uint64_t *seed) {
+	uint64_t r;
+
+	r = (*seed += 0x9e3779b97f4a7c15ULL);
+	r = (r ^ (r >> 30)) * 0xbf58476d1ce4e5b9ULL;
+	r = (r ^ (r >> 27)) * 0x94d049bb133111ebULL;
+	return (r ^ (r >> 31));
+}
+
+static inline uint64_t rotl(const uint64_t x, int k) {
+	return (x << k) | (x >> (64 - k));
+}
+
+static inline uint64_t generate_state(php_random_status_state_xoshiro256starstar *s)
+{
+    const uint64_t r = rotl(s->state[1] * 5, 7) * 9;
+	const uint64_t t = s->state[1] << 17;
+
+	s->state[2] ^= s->state[0];
+	s->state[3] ^= s->state[1];
+	s->state[1] ^= s->state[2];
+	s->state[0] ^= s->state[3];
+
+	s->state[2] ^= t;
+
+	s->state[3] = rotl(s->state[3], 45);
+
+	return r;
+}
+
+static inline void jump(php_random_status_state_xoshiro256starstar *state, const uint64_t *jmp)
+{
+	uint64_t s0 = 0, s1 = 0, s2 = 0, s3 = 0;
+	uint32_t i, j;
+
+	for (i = 0; i < 4; i++) {
+		for (j = 0; j < 64; j++) {
+			if (jmp[i] & 1ULL << j) {
+				s0 ^= state->state[0];
+				s1 ^= state->state[1];
+				s2 ^= state->state[2];
+				s3 ^= state->state[3];
+			}
+            
+			generate_state(state);
+		}
+	}
+
+	state->state[0] = s0;
+	state->state[1] = s1;
+	state->state[2] = s2;
+	state->state[3] = s3;
+}
+
+static inline void seed256(php_random_status *status, uint64_t s0, uint64_t s1, uint64_t s2, uint64_t s3)
+{
+    php_random_status_state_xoshiro256starstar *s = status->state;
+
+    s->state[0] = s0;
+    s->state[1] = s1;
+    s->state[2] = s2;
+    s->state[3] = s3;
+}
+
+static void seed(php_random_status *status, uint64_t seed)
+{
+	uint64_t s[4];
+	
+	s[0] = splitmix64(&seed);
+	s[1] = splitmix64(&seed);
+	s[2] = splitmix64(&seed);
+	s[3] = splitmix64(&seed);
+
+    seed256(status, s[0], s[1], s[2], s[3]);
+}
+
+static uint64_t generate(php_random_status *status)
+{
+    return generate_state(status->state);
+}
+
+static zend_long range(php_random_status *status, zend_long min, zend_long max)
+{
+    return php_random_range(&php_random_algo_xoshiro256starstar, status, min, max);
+}
+
+static bool serialize(php_random_status *status, HashTable *data)
+{
+    php_random_status_state_xoshiro256starstar *s = status->state;
+    zval z;
+	uint32_t i;
+
+	for (i = 0; i < 4; i++) {
+		ZVAL_STR(&z, zend_strpprintf(0, "%" PRIu64, s->state[i]));
```

Comment:
I think it would be useful if the serialized value would also be a little-endian hexadecimal string.

I'm currently working on a pure PHP polyfill for Xoshiro256**, but it is not easily possible to reproduce the serialized output, because PHP does not support unsigned integers.

I also believe that hexadecimal output is more natural for the `__debugInfo()` result, which reuses the serialization format.

Reply:
@TimWolla 
How about this? fb8efa10d643e3ce807892b90f977cb4bb44bf59

Reply:
Thanks, this is great. I have some additional remarks regarding this, but will do so in a separate discussion thread.